### PR TITLE
UI element fixes and tweaks

### LINKED ISSUE #5676: Editbox selection highlight pixel location jumps around
While changing the text selection in an edit box with <kbd>Shift</kbd> + Arrow Keys, the pixel position of the selection highlight moves around. Look at the right edge of the selection marker in the following screencast:

https://user-images.githubusercontent.com/48293446/204253208-917791bc-1f0f-45db-9def-70e76f106166.mp4

The video was made with 1.2~git25941 (edb1f99@HEAD) but it also happens in current master.

### PR DESCRIPTION
**Pull request by tothxa**
* Allow using PageUp/PageDown and Ctrl+scroll in sliders and spinboxes to change the value faster
 * Make sure that no slider nor spinbox can wrap around due to integer overflow
 * Add missing changed() call in spinbox when value is set to min/max (fixes warning update in Editor/Map Options/Ferry Range)
 * Decrease width of big step buttons of spinboxes (should help dealing with #5631)
 * Fix updating slider position when min/max is changed (affected attack window)
 * Add "Click to sort/reverse" to table header tooltips (requested by @frankystone for map selection)
 * Reset sorting order when user changes column to sort by
 * Handle big step in inputqueue
 * Allow using Shift + arrow keys in priority slider to set all input priorities at once
 * Stabilize selection end in editbox (fixes #5676)
 * Handle big step in soldier capacity settings
 * Converted all users of get_keyboard_change() to handle big step, remove enable_big_step option

### CODE REVIEW

File: src/ui_basic/table.cc [hunk 1]
```diff
@@ -145,7 +147,20 @@ void Table<void*>::set_column_tooltip(uint8_t col, const std::string& text) {
...
+
+void Table<void*>::Column::update_tooltip(bool sorted) {
+	const std::string click_tooltip =
+	   sorted ? _("Click to reverse sorting") : _("Click to sort by this column");
+	if (user_tooltip.empty()) {
+		btn->set_tooltip(click_tooltip);
+	} else {
+		/** TRANSLATORS: Tooltip format for table headers: %1 is column description, %2 is
+		    "Click to sort ..." or "Click to reverse..." */
+		btn->set_tooltip(format(_("%1s<br>(%2s)"), user_tooltip, click_tooltip));
```
Comment (Noordfrees) [inline src/ui_basic/table.cc]: Shouldn't that be:
```suggestion
		btn->set_tooltip(format(_("%1$s<br>(%2$s)"), user_tooltip, click_tooltip));
```

Comment (tothxa) [reply src/ui_basic/table.cc]: Oops. Thank you. :) And oops again, I forgot it was a suggestion.

**Comment by tothxa** [comment]
There's 2 more to go, though I may give up on #5605. And of course I may come across some more minor issues while fixing the planned ones.

**Comment by tothxa** [comment]
I'm done with all I wanted to do here. If #5677 is decided in time, I may still add it though.

File: src/wui/constructionsitewindow.cc [hunk 2]
```diff
@@ -537,25 +537,34 @@ bool ConstructionSoldierCapacityBox::handle_key(bool down, SDL_Keysym code) {
...
 	}
 	int32_t change = get_mousewheel_change(MousewheelHandlerConfigID::kChangeValue, x, y, modstate);
 	if (change == 0) {
-		return false;
+		// Try big step
+		change = get_mousewheel_change(MousewheelHandlerConfigID::kChangeValueBig, x, y, modstate);
+		if (change == 0) {
+			return false;
+		}
+		change *= 3;
```
Comment (Noordfrees) [inline src/wui/constructionsitewindow.cc]: You use the constant 3 in multiple places for big step. Please pull out into a `constexpr` variable in some header file (maybe panel.h).

File: src/ui_basic/spinbox.cc [hunk 3]
```diff
@@ -102,6 +108,13 @@ SpinBox::SpinBox(Panel* const parent,
 	} else {
 		sbi_->min = minval;
 		sbi_->max = maxval;
+#ifndef NDEBUG
+		// Prevent integer overflows
+		// 10 is the maximum value that calculate_big_step() can set
```
Comment (Noordfrees) [inline src/ui_basic/spinbox.cc]: Then 10 should be a `constexpr` variable

File: src/wui/inputqueuedisplay.cc [hunk 4]
```diff
@@ -394,33 +394,66 @@ bool InputQueueDisplay::handle_mousewheel(int32_t x, int32_t y, uint16_t modstat
...
+		big_step = true;
+	}
+
+	if (get_mouse_position().x < priority_.get_x() - kButtonSize / 4) {
+		// Mouse is over desired fill
+
+		if (big_step) {
+			// Big step of 3 is arbitrary.
+			// TODO(tothxa): Should and could it be adjusted to max amount required for a
+			//               production step?
```
Comment (Noordfrees) [inline src/wui/inputqueuedisplay.cc]: Max amount for a production step is not well-defined (what's the max amount of iron or coal needed in a barbarian war mill? Or fish in a training camp?) so let's not.

**Review by Noordfrees — COMMENTED** [review]
Lots of magic numbers, otherwise LGTM, not tested yet.

File: src/wui/constructionsitewindow.cc (hunk 2 above)
Comment (tothxa) [reply src/wui/constructionsitewindow.cc]: Done, but I'm not sure I've found the best possible solution. (I'm never sure about the right scope for shared `constexpr`s.) I put them in a namespace in `wlapplication_options.h`, next to `get_keyboard_change()`. I first thought I'd try an `enum class` (not sure whether that can act like `constexpr`), but that would need `static_cast`s all the time… yuck. :(

File: src/wui/inputqueuedisplay.cc (hunk 4 above)
Comment (tothxa) [reply src/wui/inputqueuedisplay.cc]: It's well defined, but hard to get here. :) (warmill: coal:2, iron:2, gold:2, but helmsmithy: coal:1, iron:2, gold:1; (bar) trainingcamp/fish: 2)

**Comment by Noordfrees** [comment]
LGTM now, but while testing I get this after clicking on a militarysite:
```c++
=================================================================
==18304==ERROR: AddressSanitizer: heap-buffer-overflow on address 0x6020013de174 at pc 0x562423ccd88e bp 0x7ffcefdbe900 sp 0x7ffcefdbe8f0
READ of size 4 at 0x6020013de174 thread T0
    #0 0x562423ccd88d in SoldierPanel::think() .../src/wui/soldierlist.cc:211
    #1 0x562423757655 in UI::Panel::do_think() .../src/ui_basic/panel.cc:789
    #2 0x5624237576cf in UI::Panel::do_think() .../src/ui_basic/panel.cc:805
    #3 0x5624237576cf in UI::Panel::do_think() .../src/ui_basic/panel.cc:805
    #4 0x5624237576cf in UI::Panel::do_think() .../src/ui_basic/panel.cc:805
    #5 0x5624237576cf in UI::Panel::do_think() .../src/ui_basic/panel.cc:805
    #6 0x5624237576cf in UI::Panel::do_think() .../src/ui_basic/panel.cc:805
    #7 0x5624237537ff in UI::Panel::do_run() .../src/ui_basic/panel.cc:401
    #8 0x562422ffb6e3 in UI::Panel::Returncodes UI::Panel::run<UI::Panel::Returncodes>() .../src/ui_basic/panel.h:125
    #9 0x562423430f90 in Widelands::Game::run(Widelands::Game::StartGameType, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) .../src/logic/game.cc:779
    #10 0x5624239f4e0e in FsMenu::LaunchSPG::clicked_ok() .../src/ui_fsmenu/launch_spg.cc:173
    #11 0x5624238449e2 in FsMenu::TwoColumnsFullNavigationMenu::handle_key(bool, SDL_Keysym) .../src/ui_fsmenu/menu.cc:161
    #12 0x56242375c7fa in UI::Panel::do_key(bool, SDL_Keysym) .../src/ui_basic/panel.cc:1407
    #13 0x56242375c6bf in UI::Panel::do_key(bool, SDL_Keysym) .../src/ui_basic/panel.cc:1399
    #14 0x56242375c6bf in UI::Panel::do_key(bool, SDL_Keysym) .../src/ui_basic/panel.cc:1399
    #15 0x56242375e0cf in UI::Panel::ui_key(bool, SDL_Keysym) .../src/ui_basic/panel.cc:1653
    #16 0x562422fd6671 in WLApplication::handle_input(InputCallback const*) .../src/wlapplication.cc:956
    #17 0x5624237535d1 in UI::Panel::do_run() .../src/ui_basic/panel.cc:389
    #18 0x562422ff8f31 in int UI::Panel::run<int>() .../src/ui_basic/panel.h:125
    #19 0x562422fd36cc in WLApplication::run() .../src/wlapplication.cc:835
    #20 0x562422fc6e6a in main .../src/main.cc:45
    #21 0x7fb21fa2350f in __libc_start_call_main ../sysdeps/nptl/libc_start_call_main.h:58
    #22 0x7fb21fa235c8 in __libc_start_main_impl ../csu/libc-start.c:381
    #23 0x562422fc6c84 in _start (.../widelands+0x61ac84)

0x6020013de174 is located 0 bytes to the right of 4-byte region [0x6020013de170,0x6020013de174)
allocated by thread T0 here:
    #0 0x7fb2226c0488 in operator new(unsigned long) ../../../../src/libsanitizer/asan/asan_new_delete.cpp:95
    #1 0x562423454f07 in std::__new_allocator<unsigned int>::allocate(unsigned long, void const*) /usr/include/c++/12/bits/new_allocator.h:137
    #2 0x56242345191f in std::allocator_traits<std::allocator<unsigned int> >::allocate(std::allocator<unsigned int>&, unsigned long) /usr/include/c++/12/bits/alloc_traits.h:464
    #3 0x56242344c563 in std::_Vector_base<unsigned int, std::allocator<unsigned int> >::_M_allocate(unsigned long) /usr/include/c++/12/bits/stl_vector.h:378
    #4 0x56242344a4ad in std::vector<unsigned int, std::allocator<unsigned int> >::_M_default_append(unsigned long) /usr/include/c++/12/bits/vector.tcc:650
    #5 0x5624234451d8 in std::vector<unsigned int, std::allocator<unsigned int> >::resize(unsigned long) /usr/include/c++/12/bits/stl_vector.h:1011
    #6 0x562423ccd38b in SoldierPanel::think() .../src/wui/soldierlist.cc:188
    #7 0x562423757655 in UI::Panel::do_think() .../src/ui_basic/panel.cc:789
    #8 0x5624237576cf in UI::Panel::do_think() .../src/ui_basic/panel.cc:805
    #9 0x5624237576cf in UI::Panel::do_think() .../src/ui_basic/panel.cc:805
    #10 0x5624237576cf in UI::Panel::do_think() .../src/ui_basic/panel.cc:805
    #11 0x5624237576cf in UI::Panel::do_think() .../src/ui_basic/panel.cc:805
    #12 0x5624237576cf in UI::Panel::do_think() .../src/ui_basic/panel.cc:805
    #13 0x5624237537ff in UI::Panel::do_run() .../src/ui_basic/panel.cc:401
    #14 0x562422ffb6e3 in UI::Panel::Returncodes UI::Panel::run<UI::Panel::Returncodes>() .../src/ui_basic/panel.h:125
    #15 0x562423430f90 in Widelands::Game::run(Widelands::Game::StartGameType, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) .../src/logic/game.cc:779
    #16 0x5624239f4e0e in FsMenu::LaunchSPG::clicked_ok() .../src/ui_fsmenu/launch_spg.cc:173
    #17 0x5624238449e2 in FsMenu::TwoColumnsFullNavigationMenu::handle_key(bool, SDL_Keysym) .../src/ui_fsmenu/menu.cc:161
    #18 0x56242375c7fa in UI::Panel::do_key(bool, SDL_Keysym) .../src/ui_basic/panel.cc:1407
    #19 0x56242375c6bf in UI::Panel::do_key(bool, SDL_Keysym) .../src/ui_basic/panel.cc:1399
    #20 0x56242375c6bf in UI::Panel::do_key(bool, SDL_Keysym) .../src/ui_basic/panel.cc:1399
    #21 0x56242375e0cf in UI::Panel::ui_key(bool, SDL_Keysym) .../src/ui_basic/panel.cc:1653
    #22 0x562422fd6671 in WLApplication::handle_input(InputCallback const*) .../src/wlapplication.cc:956
    #23 0x5624237535d1 in UI::Panel::do_run() .../src/ui_basic/panel.cc:389
    #24 0x562422ff8f31 in int UI::Panel::run<int>() .../src/ui_basic/panel.h:125
    #25 0x562422fd36cc in WLApplication::run() .../src/wlapplication.cc:835
    #26 0x562422fc6e6a in main .../src/main.cc:45
    #27 0x7fb21fa2350f in __libc_start_call_main ../sysdeps/nptl/libc_start_call_main.h:58

SUMMARY: AddressSanitizer: heap-buffer-overflow .../src/wui/soldierlist.cc:211 in SoldierPanel::think()
Shadow bytes around the buggy address:
  0x0c0480273bd0: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c0480273be0: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c0480273bf0: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c0480273c00: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c0480273c10: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
=>0x0c0480273c20: fa fa fa fa fa fa fa fa fa fa fa fa fa fa[04]fa
  0x0c0480273c30: fa fa fa fa fa fa fa fa fa fa fd fd fa fa fa fa
  0x0c0480273c40: fa fa fa fa fa fa fd fa fa fa fa fa fa fa 00 00
  0x0c0480273c50: fa fa fd fa fa fa 00 00 fa fa fd fd fa fa 00 fa
  0x0c0480273c60: fa fa 00 fa fa fa fd fa fa fa fd fa fa fa fd fd
  0x0c0480273c70: fa fa fd fd fa fa fd fa fa fa fd fa fa fa 00 fa
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07 
  Heap left redzone:       fa
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
  Left alloca redzone:     ca
  Right alloca redzone:    cb
==18304==ABORTING
```

**Comment by tothxa** [comment]
> LGTM now, but while testing I get this after clicking on a militarysite:

Is it another integer overflow in `CmdChangeSoldierCapacity::execute()`? I think I missed before that finished buildings use the received value as a delta, only constructionsites get the new value. I guess I shouldn't unify this though, because that may cause major incompatibilities?

(edit: on closer look it's OK, the value is cast to signed for adding the (possibly negative) delta)

**Review by Noordfrees — APPROVED** [review]
At closer look that seems to be an unrelated crash: #5652

Code LGTM, all aspects tested and working :)

**Comment by tothxa** [comment]
Thank you for the review.

**Pull request merged**